transformer_engine/jax/transformer.py

```diff
@@ -0,0 +1,911 @@
+        the data type used to allocate the initial parameters.
+    fuse_qkv: bool, default = True
+        if set to True, this module exposes a single fused
+        parameter for query-key-value for self-attention and key-value for
+        cross-attention.
+    transpose_batch_sequence : bool, default = True
+        indicate whether the input tensors were switched axis of batch
+        and sequence length dimension. if set to True, the input tensors
+        should be in (seqlen, batch, hidden), otherwise (batch, seqlen, hidden).
+    scale_attn_logits: bool, default = False
+        indicate whether to scale attention logits.
+        if set to True, :math:`\frac{Q}{\sqrt{head_dim}*K}`,
+        else :math:`Q*K`
+    scaled_query_init: bool, default = `True`
+        whether to scale WQ on initilization by :math:`\sqrt{head_dim}`
+    float32_logits : bool, default = False
+        whether to compute attention logits in float32.
+    """
+    # pylint: enable=line-too-long
+    head_dim: int
+    num_heads: int
+    dropout_rate: float = 0.
+    dropout_rng_name: str = 'dropout'
+    layernorm_type: str = "layernorm"
+    layernorm_epsilon: float = 1e-6
+    kernel_init: Initializer = nn.initializers.variance_scaling(1.0, 'fan_in', 'normal')
+    use_bias: bool = False
+    bias_init: Initializer = nn.initializers.zeros
+    apply_residual_connection_post_layernorm: bool = False
+    output_layernorm: bool = False
+    attn_type: AttentionType = AttentionType.PADDING
+    dtype: DType = jnp.float32
+    fuse_qkv: bool = True
+    transpose_batch_sequence: bool = True
+    scale_attn_logits: bool = False
+    scaled_query_init: bool = True
+    float32_logits: bool = False    # computes logits in float32 for stability.
+
+    @nn.compact
+    def __call__(self,
+                 inputs_q: Array,
+                 inputs_kv: Array,
+                 mask: Optional[Array] = None,
+                 bias: Optional[Array] = None,
+                 *,
+                 decode: bool = False,
+                 deterministic: bool = False) -> Array:
+        """Applies multi-head dot product attention on the input data."""
+
+        depth_scaling = jnp.sqrt(self.head_dim).astype(self.dtype)
+
+        def query_init(*args):
+            return self.kernel_init(*args) / (depth_scaling if self.scaled_query_init else 1.0)
+
+        def qkv_init(key, shape, dtype):
+            assert len(shape) == 3
+            assert shape[-2] == 3
+
+            q_key, k_key, v_key = jax_random.split(key, num=3)
+
+            q_shape = (shape[0], shape[-1])
+            k_shape = (shape[0], shape[-1])
+            v_shape = (shape[0], shape[-1])
+
+            q_kernel = query_init(q_key, q_shape, dtype)
+            k_kernel = self.kernel_init(k_key, k_shape, dtype)    # pylint: disable=too-many-function-args
+            v_kernel = self.kernel_init(v_key, v_shape, dtype)    # pylint: disable=too-many-function-args
```

Comment:
Are these Pylint suppressions necessary? I see that [JAX initializers](https://jax.readthedocs.io/en/latest/jax.nn.initializers.html) take 3 args, so why would it complain?

Reply:
If something is expected as a kwarg, let's pass it in as such, I think that's the issue

Reply:
We don't know why it triggers `too-many-function-args`. We tried `self.kernel_init(key=k_key, shape=k_shape, dtype=dtype)` but it triggered aonther error - `redundant-keyword-arg`. I think it is an issue between jax.nn and pylint. We should focus on our effort on other places

Reply:
I wonder if it's because we're running PyLint in the PyTorch container, so it can't figure out the function signatures of JAX functions. If this is the case, suppressing these warnings is fine for now. In the future we'll want to run PyLint for these files using the JAX container.

Reply:
I think I understand why this error is happening. Pylint doesn't fully understand dataclasses and it thinks we are setting `kernel_init` as a class function instead of an instance variable. It thinks that we need to insert `self` as an arg, so obviously we have one more arg than expected. This error can be safely suppressed.

Reply:
On a similar note, having mutable or complex default args is usually not a good idea. A better method might be to set it to `None` and then assign the default value during init when `None` is seen

Reply:
@ksivaman ,
You are talking about 
```python
kernel_init: Initializer = nn.initializers.variance_scaling(1.0, 'fan_in', 'truncated_normal')
```
, right?

In flax.nn.Module, the data member cannot be changed in `__call__`, `__init__` and `setup` because module instance is frozen. The `__post_init__` is the only place that can modify data members.

The commit [4f104c1](https://github.com/NVIDIA/TransformerEngine/pull/54/commits/4f104c1a5dadb2086ea3e83d47ac88079e2e277d) moves initialing the *kernel_init to `__post_init___`.